File: src/main/c/easy/q1636/Solution.c
#include <stdio.h>
#include <stdlib.h>

#define OFFSET 100
#define RANGE 201

int cmp(const void* a, const void* b, void* freq_map) {
    int x = *(int*)a;
    int y = *(int*)b;
    int* freq = (int*)freq_map;

    if (freq[x + OFFSET] != freq[y + OFFSET])
        return freq[x + OFFSET] - freq[y + OFFSET];
    return y - x; // descending value
}

int* frequencySort(int* nums, int numsSize, int* returnSize) {
    int* freq = calloc(RANGE, sizeof(int));
    for (int i = 0; i < numsSize; i++) {
        freq[nums[i] + OFFSET]++;
    }

    qsort_r(nums, numsSize, sizeof(int), cmp, freq);

    *returnSize = numsSize;
    free(freq);
    return nums;
}

File: src/main/rust/easy/q1636/Solution.rs
use std::collections::HashMap;

impl Solution {
    pub fn frequency_sort(nums: Vec<i32>) -> Vec<i32> {
        let mut freq = HashMap::new();
        for &num in &nums {
            *freq.entry(num).or_insert(0) += 1;
        }

        let mut nums = nums;
        nums.sort_by(|a, b| {
            let fa = freq[a];
            let fb = freq[b];
            if fa == fb {
                b.cmp(a) // value descending
            } else {
                fa.cmp(&fb) // freq ascending
            }
        });

        nums
    }
}
